File: vcf_process.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Parse bcftools mpileup VCF, extract per-sample DP/AD and compute ratios.

Features:
- Preserves #CHROM column names; skips ## metadata
- Compatible with .vcf and .vcf.gz
- Automatic path normalization on Windows/WSL
- Use --input to specify the input directory; --output to specify the output directory
- Infers sample columns from headers (can also be specified with --samples)
- Prioritizes dynamic DP/AD location based on FORMAT; falls back to a fixed location if not found
"""

import os
import re
import gzip
import argparse
import platform
import pandas as pd
import numpy as np
from typing import List, Optional
import tempfile
import shutil

# ---------- Path Normalization (Windows -> WSL) ----------
def normalize_path(p: Optional[str]) -> Optional[str]:
    if p is None:
        return p
    p = p.strip().strip('"').strip("'")
    # Map "E:\path" or "e:/path" to "/mnt/e/path" under Linux/WSL
    if platform.system() == "Linux":
        m = re.match(r'^([a-zA-Z]):[\\/](.*)$', p)
        if m:
            drive = m.group(1).lower()
            rest = m.group(2).replace('\\', '/')
            return f"/mnt/{drive}/{rest}"
    return p.replace('\\', '/')

# ---------- Security Split ----------
def safe_split(s: Optional[str], sep: str, idx: int) -> Optional[str]:
    if s is None or (isinstance(s, float) and pd.isna(s)):
        return np.nan
    parts = str(s).split(sep)
    return parts[idx] if 0 <= idx < len(parts) else np.nan

def find_column_index(fmt: str, key: str) -> Optional[int]:
    """Find the 0-based position of the key in the FORMAT field"""
    if fmt is None or (isinstance(fmt, float) and pd.isna(fmt)):
        return None
    fields = str(fmt).split(':')
    for i, k in enumerate(fields):
        if k == key:
            return i
    return None

# ---------- Sample name cleaning (remove paths, remove .bam/.cram, and remove illegal characters -> _) ----------
def sanitize_sample(name: str) -> str:
    base = os.path.basename(name)
    base = re.sub(r"\.(bam|cram)$", "", base, flags=re.IGNORECASE)
    base = re.sub(r"[^\w.\-]+", "_", base)
    return base

# ---------- parameter ----------
def parse_args():
    ap = argparse.ArgumentParser(description="Fix strand and compute mutation ratios from VCF.")
    ap.add_argument("s", type=int, help="strand: 1=FWD/neg, 2=REV/pos")
    ap.add_argument("--input", required=True, help="Input VCF directory (including ProQ-rABE_FWD/REV.vcf[.gz])")
    ap.add_argument("--output", required=True, help="Output Directory")
    ap.add_argument("--samples", nargs="*", default=None,
                    help="Optional: list of sample names; if not provided, they will be automatically inferred from the VCF header")
    return ap.parse_args()

# ---------- Main process ----------
def main():
    args = parse_args()

    # strand
    if args.s == 1:
        st, strand = "FWD", "neg"
    elif args.s == 2:
        st, strand = "REV", "pos"
    else:
        raise SystemExit(" s can be only 1 or 2.")

    input = normalize_path(args.input)
    output  = normalize_path(args.output)
    os.makedirs(output, exist_ok=True)

    # Input file (prefer .vcf, if not available, try .vcf.gz)
    vcf_path = os.path.join(input, f"ProQ-rABE_{st}.vcf")
    vcf_gz   = vcf_path + ".gz"
    if not os.path.exists(vcf_path):
        if os.path.exists(vcf_gz):
            vcf_path = vcf_gz
        else:
            raise FileNotFoundError(f"Input file not found: \n {vcf_path}\n or {vcf_gz}")

    print(f"[INFO] read file: {vcf_path}")

    # Read the header (#CHROM line) to get the column names and sample columns
    open_func = gzip.open if vcf_path.endswith(".gz") else open
    with open_func(vcf_path, "rt") as f:
        header_line = None
        for line in f:
            if line.startswith("#CHROM"):
                header_line = line.strip().lstrip("#").split("\t")
                break
    if header_line is None:
        raise ValueError("No #CHROM header row found in VCF.")

    # The sample columns start from the 10th column of the header.
    sample_cols_from_header = header_line[9:]

    # Read data: skip lines starting with # and use header_line as column names
    df = pd.read_csv(
        vcf_path,
        sep="\t",
        comment="#",
        names=header_line,
        dtype=str,
        na_filter=True,
        engine="python",
        compression="infer",
    )

    # Strand and ALT processing
    df["strand"] = strand
    df["ALT"] = df["ALT"].astype(str).str.slice(0, 1)

    # Drop rows with missing core columns (similar to R’s drop_na )
    core_cols = ["CHROM", "POS", "REF", "ALT", "FORMAT"]
    df = df.dropna(subset=[c for c in core_cols if c in df.columns]).copy()

    # RNAREF / RNAALT
    df["RNAREF"] = df["REF"].astype(str)
    df["RNAALT"] = df["ALT"].astype(str)
    comp = {"A": "T", "T": "A", "C": "G", "G": "C"}
    if strand == "neg":
        df["RNAREF"] = df["RNAREF"].map(comp).fillna(df["RNAREF"])
        df["RNAALT"] = df["RNAALT"].map(comp).fillna(df["RNAALT"])

    # Sample name: User-provided is preferred, otherwise inferred from the header and cleaned
    if args.samples and len(args.samples) > 0:
        if len(args.samples) != len(sample_cols_from_header):
            print("[WARN] The number of --samples is inconsistent with the number of VCF sample columns. The VCF header will take precedence.")
            sList: List[str] = [sanitize_sample(x) for x in sample_cols_from_header]
        else:
            sList = args.samples
    else:
        sList = [sanitize_sample(x) for x in sample_cols_from_header]

    sample_cols = sample_cols_from_header  # 与 df 中样本列一一对应

    print("[INFO] Sample columns (raw):" + ", ".join(sample_cols))
    print("[INFO] Sample names (for column prefixes):" + ", ".join(sList))

    # Parse DP/AD for each sample by FORMAT (fallback to fixed position)
    def extract_sample_dp_ad(sample_series: pd.Series, format_series: pd.Series):
        dp_vals, ad_vals = [], []
        for fmt, cell in zip(format_series, sample_series):
            dp_idx = find_column_index(fmt, "DP")
            ad_idx = find_column_index(fmt, "AD")
            if dp_idx is None: dp_idx = 1   # Fallback: Paragraph 2  # noqa: E701
            if ad_idx is None: ad_idx = 5   # Fallback: Paragraph 6  # noqa: E701
            dp_str = safe_split(cell, ":", dp_idx)
            ad_str = safe_split(cell, ":", ad_idx)
            alt1  = safe_split(ad_str, ",", 1) if pd.notna(ad_str) else np.nan
            dp_vals.append(pd.to_numeric(dp_str, errors="coerce"))
            ad_vals.append(pd.to_numeric(alt1,  errors="coerce"))
        return pd.Series(dp_vals), pd.Series(ad_vals)

    for sample_col, sp in zip(sample_cols, sList):
        dp_s, ad_s = extract_sample_dp_ad(df[sample_col], df["FORMAT"])
        df[f"{sp}.DP"] = dp_s
        df[f"{sp}.AD"] = ad_s
        with np.errstate(divide="ignore", invalid="ignore"):
            df[f"{sp}.ratio"] = (df[f"{sp}.AD"].astype(float) /
                                 df[f"{sp}.DP"].astype(float)).replace([np.inf, -np.inf], np.nan)

    # Average coverage (DP average)
    dp_cols = [f"{sp}.DP" for sp in sList]
    df["avgDP"] = df[dp_cols].mean(axis=1, skipna=True)

# === Write out the file (Windows writes directly; Linux/WSL writes to /tmp first and then copies) ===


    out_path = os.path.join(output, f"mpileup_fixstrand_{strand}.vcf")
    out_path = os.path.normpath(out_path)
    print("[INFO] Target output path repr:", repr(out_path))
    os.makedirs(os.path.dirname(out_path), exist_ok=True)

    if platform.system() == "Windows":
        # Windows Write the target path directly
        df.to_csv(out_path, sep="\t", index=False, header=True)
        print(f"[OK] Output File: {out_path}  (total {len(df)} lines）")
    else:
        # Linux/WSL：Write to /tmp first, then copy to /mnt/disk to avoid drvfs Errno 22
        with tempfile.NamedTemporaryFile(mode="w", encoding="utf-8", delete=False,
                                        dir="/tmp", prefix="vcf_", suffix=".vcf") as tf:
            tmp_out = tf.name
        df.to_csv(tmp_out, sep="\t", index=False, header=True)
        try:
            shutil.copyfile(tmp_out, out_path)
            print(f"[OK] Output File: {out_path}  (total {len(df)} lines）")
        except OSError as e:
            print("[WARN] Copying to the target directory failed:", repr(e))
            print(f"[HINT] Please copy manually: cp {tmp_out} {out_path}")

if __name__ == "__main__":
    main()

File: vcf_process.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(tidyverse)   # dplyr / stringr / readr / tidyr
})

# ---------------- 参数解析（无需额外依赖） ----------------
parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  get_val <- function(key, default = NULL) {
    hit_eq <- grep(paste0("^--", key, "="), args, value = TRUE)
    if (length(hit_eq) > 0) return(sub(paste0("^--", key, "="), "", hit_eq[1]))
    hit <- which(args == paste0("--", key))
    if (length(hit) > 0 && length(args) >= hit + 1) return(args[hit + 1])
    default
  }
  # s 既支持位置参数也支持 --s
  s_val <- get_val("s", NA)
  if (is.na(s_val)) {
    if (length(args) >= 1 && !startsWith(args[1], "--")) s_val <- args[1]
  }
  list(
    s        = suppressWarnings(as.integer(ifelse(is.na(s_val), 1L, s_val))),
    workdir  = get_val("workdir", "."),
    resdir   = get_val("resdir",  "."),
    samples  = get_val("samples", NULL)  # 逗号分隔，可选
  )
}

# ---------------- 读取 VCF 表头 (#CHROM) ----------------
read_vcf_header <- function(path) {
  con <- if (grepl("\\.gz$", path, ignore.case = TRUE)) gzfile(path, "rt") else file(path, "rt")
  on.exit(close(con), add = TRUE)
  header <- NULL
  repeat {
    line <- readLines(con, n = 1, warn = FALSE)
    if (length(line) == 0) break
    if (startsWith(line, "#CHROM")) {
      header <- strsplit(sub("^#", "", line), "\t", fixed = FALSE)[[1]]
      break
    }
  }
  header
}

# ---------------- 样本名清洗 ----------------
sanitize_sample <- function(x) {
  b <- basename(x)
  b <- sub("\\.(bam|cram)$", "", b, ignore.case = TRUE)
  b <- gsub("[^[:alnum:]_.-]+", "_", b)
  b
}

# ---------------- 主流程 ----------------
main <- function() {
  a <- parse_args()
  if (!(a$s %in% c(1L, 2L))) stop("参数 s 只能是 1 或 2（1=FWD/neg, 2=REV/pos）。")

  st     <- if (a$s == 1L) "FWD" else "REV"
  strand <- if (a$s == 1L) "neg" else "pos"

  workdir <- a$workdir
  resdir  <- a$resdir
  if (!dir.exists(resdir)) dir.create(resdir, recursive = TRUE, showWarnings = FALSE)

  vcf_path <- file.path(workdir, sprintf("ProQ-rABE_%s.vcf", st))
  if (!file.exists(vcf_path) && file.exists(paste0(vcf_path, ".gz"))) vcf_path <- paste0(vcf_path, ".gz")
  if (!file.exists(vcf_path)) stop("未找到输入：", vcf_path, " 或其 .gz 版本")

  message("[INFO] 读取文件: ", vcf_path)

  header <- read_vcf_header(vcf_path)
  if (is.null(header)) stop("未在 VCF 中找到 #CHROM 表头行。")
  sample_cols <- header[10:length(header)]
  if (length(sample_cols) == 0) stop("VCF 中未检测到样本列。")

  # 读取数据（跳过 # 行），全部按字符读入
  df <- readr::read_tsv(
    vcf_path,
    comment   = "#",
    col_names = header,
    col_types = readr::cols(.default = readr::col_character()),
    progress  = FALSE
  )

  # 新增 strand；ALT 仅保留首个碱基；去除核心缺失
  df <- df %>%
    mutate(
      strand = strand,
      ALT    = substr(ALT, 1, 1)
    ) %>%
    drop_na(CHROM, POS, REF, ALT)

  # RNAREF / RNAALT（负链互补）—— 使用“向量安全”的方式
  comp <- c(A = "T", T = "A", C = "G", G = "C")
  df <- df %>%
    mutate(
      RNAREF = REF,
      RNAALT = ALT
    )
  if (strand == "neg") {
    # 先映射，返回可能带 NA 的向量；再用 ifelse 替换 NA 为原值（向量安全）
    rnaref_map <- unname(comp[df$RNAREF])
    rnaalt_map <- unname(comp[df$RNAALT])
    df$RNAREF  <- ifelse(is.na(rnaref_map), df$RNAREF, rnaref_map)
    df$RNAALT  <- ifelse(is.na(rnaalt_map), df$RNAALT, rnaalt_map)
  }

  # 样本名（优先 --samples；否则从 VCF 表头推断并清洗）
  if (!is.null(a$samples) && nchar(a$samples) > 0) {
    user_samples <- strsplit(a$samples, ",", fixed = TRUE)[[1]] %>% trimws()
    if (length(user_samples) != length(sample_cols)) {
      warning("--samples 数量与 VCF 样本列不一致，将使用 VCF 表头推断。")
      sList <- sanitize_sample(sample_cols)
    } else {
      sList <- user_samples
    }
  } else {
    sList <- sanitize_sample(sample_cols)
  }

  message("[INFO] 样本列（VCF头）：", paste(sample_cols, collapse = ", "))
  message("[INFO] 样本名（列前缀）：", paste(sList, collapse = ", "))

  # ------- 固定位置解析：DP=第二段, AD=第六段；AD 取逗号第二项（alt） -------
  # 向量安全：使用 str_split_fixed 取到第 6 段，再用 strsplit() 提取“第二个逗号段”
  for (k in seq_along(sample_cols)) {
    col <- sample_cols[k]
    sp  <- sList[k]

    parts  <- stringr::str_split_fixed(df[[col]], ":", 6)
    DP_chr <- parts[, 2]
    AD_raw <- parts[, 6]

    AD_alt_chr <- vapply(
      strsplit(AD_raw, ",", fixed = TRUE),
      FUN = function(x) if (length(x) >= 2) x[2] else NA_character_,
      FUN.VALUE = character(1)
    )

    df[[paste0(sp, ".DP")]]    <- suppressWarnings(as.numeric(DP_chr))
    df[[paste0(sp, ".AD")]]    <- suppressWarnings(as.numeric(AD_alt_chr))
    # ratio
    num <- df[[paste0(sp, ".AD")]]
    den <- df[[paste0(sp, ".DP")]]
    ratio <- num / den
    ratio[is.infinite(ratio)] <- NA_real_
    df[[paste0(sp, ".ratio")]] <- ratio
  }

  # 平均覆盖
  dp_cols   <- paste0(sList, ".DP")
  df$avgDP  <- rowMeans(as.matrix(df[dp_cols]), na.rm = TRUE)

  # 输出
  out_path <- file.path(resdir, sprintf("mpileup_fixstrand_%s.vcf", strand))
  message("[INFO] 输出路径: ", out_path)

  ok <- tryCatch({
    write.table(df, file = out_path, sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)
    TRUE
  }, error = function(e) {
    message("[WARN] 直接写出失败：", conditionMessage(e))
    FALSE
  })

  if (!ok) {
    tmp <- tempfile(fileext = ".vcf")
    write.table(df, file = tmp, sep = "\t", quote = FALSE, row.names = FALSE, col.names = TRUE)
    message("[HINT] 已写入临时文件：", tmp)
    message("[HINT] 请手动复制到目标目录，例如：")
    message("       cp ", shQuote(tmp), " ", shQuote(out_path))
  } else {
    message("[OK] 完成：", out_path, "  （共 ", nrow(df), " 行）")
  }
}

main()
